src/sentry/eventstream/kafka/postprocessworker.py

```diff
@@ -0,0 +1,134 @@
+
+
+def _record_metrics(partition: int, task_kwargs: Mapping[str, Any]) -> None:
+    event_type = "transactions" if task_kwargs["group_id"] is None else "errors"
+    metrics.incr(
+        "eventstream.messages",
+        tags={"partition": partition, "type": event_type},
+    )
+
+
+def dispatch_post_process_group_task(
+    event_id: str,
+    project_id: int,
+    group_id: Optional[int],
+    is_new: bool,
+    is_regression: bool,
+    is_new_group_environment: bool,
+    primary_hash: Optional[str],
+    skip_consume: bool = False,
+) -> None:
+    if skip_consume:
+        logger.info("post_process.skip.raw_event", extra={"event_id": event_id})
+    else:
+        cache_key = cache_key_for_event({"project": project_id, "event_id": event_id})
+
+        post_process_group.delay(
+            is_new=is_new,
+            is_regression=is_regression,
+            is_new_group_environment=is_new_group_environment,
+            primary_hash=primary_hash,
+            cache_key=cache_key,
+            group_id=group_id,
+        )
+
+
+def _get_task_kwargs_and_dispatch(message: Message):
+    task_kwargs = _get_task_kwargs(message)
+    if not task_kwargs:
+        return None
+
+    _record_metrics(message.partition(), task_kwargs)
+    dispatch_post_process_group_task(**task_kwargs)
+
+
+class PostProcessForwarderWorker(AbstractBatchWorker):
+    """
+    Implementation of the AbstractBatchWorker which would be used for post process forwarder.
+    The current implementation creates a thread pool worker based on the concurrency parameter
+    because we want to be able to change the concurrency during runtime. This should be replaced
+    by a thread pool executor once stress tests experiments are over and we start using the
+    CLI arguments to set concurrency.
+    """
+
+    def __init__(self, concurrency: Optional[int] = 1) -> None:
+        self.__current_concurrency = concurrency
+        self.__executor = ThreadPoolExecutor(max_workers=self.__current_concurrency)
+
+    def process_message(self, message: Message) -> Optional[Future]:
+        """
+        Process the message received by the consumer and return the Future associated with the message. The future
+        is stored in the batch of batching_kafka_consumer and provided as an argument to flush_batch. If None is
+        returned, the batching_kafka_consumer will not add the return value to the batch.
+        """
+        return self.__executor.submit(_get_task_kwargs_and_dispatch, message)
+
+    def flush_batch(self, batch: Sequence[Future]) -> None:
+        """
+        For all work which was submitted to the thread pool executor, we need to ensure that if an exception was
+        raised, then we raise it in the main thread. This is needed so that processing can be stopped in such
+        cases.
+        """
+        for future in as_completed(batch):
+            exc = future.exception()
+            if exc is not None:
+                raise exc
```

Comment:
If this fails in between a batch we would we should not process and execute again the messages that we already successfully scheduled. 
Now I am not 100% sure if, when raising the exception here, all partitions will be revoked from this consumer. If that is the case, and you commit the staged offset in the on_revoke method, you will achieve what you need by successfully committing the offsets that were successfully scheduled. 
If not we will need to do that some other way. 
I think the on_revoke is called in that case, but could you please verify ?

Reply:
This is not a requirement for the post_process_group task. It works with at least once semantics. There is a check for the cached event being present in the event processing store. If the event is not present it means its already been handled and won't do duplicate work. The code can be found here https://github.com/getsentry/sentry/blob/master/src/sentry/tasks/post_process.py#L206-L215

If we need exactly once guarantees from the consumer, then it would be a little tricky since this uses a threadpool and the ordering of work is not guaranteed. Let me know if we need this to be a consumer requirement.

Reply:
Actually nevermind. It is already providing an "at least one" guarantee. If any message fails to be processed we raise an exception and that's it. I don't think the on_revoke method will be called (mind testing this) so the uncommitted messages would be reprocessed